vpr/src/place/noc_place_utils.h

```diff
@@ -0,0 +1,375 @@
+
+/**
+ * @brief Goes through all the traffic flows and determines whether the
+ * latency constraints have been met for each traffic flow. 
+ * 
+ * @return The total number of traffic flows with latency constraints being met
+ */
+int get_number_of_traffic_flows_with_latency_cons_met(void);
+
+/**
+ * @brief THis verifies whether we need to compare the recomputed noc latency
+ * cost to the current noc latency cost. During placement there is a function
+ * called "recompute_costs_from_scratch" that regularily checks whether the
+ * NoC latency cost updates after each move match a complete recalculation of
+ * the NoC latency cost at the current state. This will fail when the latency
+ * costs get really low. So we expect the latency costs to never actually go 
+ * below one picosecond, so we check whether a recomputed cost is below this
+ * threshold and if it is then indicate that there is no need to verify the
+ * NoC latency cost. This is acceptable since the only case where the NoC
+ * latency cost will be below one picosecond is when the weighting on the 
+ * latency component of the cost is extremely low (ie. 0) and when most if not 
+ * all the latency constraints have been met (so this is also close to 0). There
+ * is no need to check and compare the cost here.
+ * 
+ */
+bool check_recomputed_noc_latency_cost(float recomputed_cost);
+
+void allocate_and_load_noc_placement_structs(void);
+
+void free_noc_placement_structs(void);
+
+/* Below are functions related to the feature that forces to the placer to swap router blocks for a certain percentage of the total number of swaps */
+
+/**
+ * @brief User supplied a number that represents the percentage of router 
+ * block swaps relative to the total number of swaps. A random number  
+ * between 0-100 is generated here and compared the user supplied value.
+ * If the random number is less than the user supplied value we indicate
+ * that a router block should be swapped. By doing this we now only swap
+ * router blocks for the percentage of time the user supplied.
+ * 
+ * @param user_supplied_noc_router_swap_percentage The expected number of 
+ * noc router block swaps as a percentage of the total number of swaps
+ * attempted by the placer.
+ * @return true This indicates that a router block should be swapped
+ * @return false This indicates that a router block should not be swapped
+ */
+bool check_for_router_swap(int user_supplied_noc_router_swap_percentage);
+
+/**
+ * @brief Generates a placement move by choosing two rouer cluster blocks to 
+ * swap. First, a random router cluster block is chosen and then another router
+ * cluster block is chosen that cqan be swapped with the initial block such that
+ * the distance travelled by either blocks does not exceed rlim. 
+ * 
+ * @param blocks_affected The two router cluster blocks that are proposed to be
+ * swapped
+ * @param rlim The maximum distance in the x and y direction that a router 
```

Comment:
x and y in the compressed block space?  (i.e. rlim == 1 means swap only with next router to your left or right)

Reply:
Yes, this is within the compressed block space. Comment updated